### PR TITLE
Add ARM MUSL targets

### PR DESCRIPTION
**Pull request by japaric**
Rebase of #33189.

I tested this by producing a std for `arm-unknown-linux-musleabi` then I cross compiled Hello world to said target. Checked that the produced binary was statically linked and verified that the binary worked under QEMU.

This depends on rust-lang/libc#341. I'll have to update this PR after that libc PR is merged.

I'm also working on generating ARM musl cross toolchain via crosstool-ng. Once I verified those work, I'll send a PR to rust-buildbot.

r? @alexcrichton 
cc @timonvo

### CODE REVIEW

File: src/librustc_back/target/musl_base.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
```
Comment (eefriedman) [inline src/librustc_back/target/musl_base.rs]: Is this an exact copy of src/librustc_back/target/linux_musl_base.rs?

Comment (japaric) [inline src/librustc_back/target/musl_base.rs]: Indeed, it is! I'll remove it.

File: src/libstd/rtdeps.rs [hunk 2]
```diff
@@ -19,7 +19,7 @@
 //
 // On Linux, librt and libdl are indirect dependencies via std,
 // and binutils 2.22+ won't add them automatically
-#[cfg(all(target_os = "linux", not(target_env = "musl")))]
+#[cfg(all(target_os = "linux", not(any(target_env = "musl"))))]
```
Comment (alexcrichton) [inline src/libstd/rtdeps.rs]: The addition of `any(...)` here and below may no longer be necessary

**Comment by alexcrichton** [comment]
Looks great to me, thanks @japaric! I'll r+ once the libc PR has merged

**Comment by japaric** [comment]
Addressed comment about any and updated the libc submodule.

@bors: r=alexcrichton

**Comment by japaric** [comment]
> Once I verified those work, I'll send a PR to rust-buildbot.

See rust-lang/rust-buildbot#109

**Comment by alexcrichton** [comment]
@bors: rollup

assuming this isn't very platform-specific and green travis means it's good to go whenever

File: src/librustc_back/target/arm_unknown_linux_musleabi.rs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use target::Target;
+
+pub fn target() -> Target {
```
Comment (cardoe) [inline src/librustc_back/target/arm_unknown_linux_musleabi.rs]: This needs to return `TargetResult` now.

File: src/librustc_back/target/arm_unknown_linux_musleabihf.rs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use target::Target;
+
+pub fn target() -> Target {
```
Comment (cardoe) [inline src/librustc_back/target/arm_unknown_linux_musleabihf.rs]: This needs to return `TargetResult` now.

File: src/librustc_back/target/armv7_unknown_linux_musleabihf.rs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use target::Target;
+
+pub fn target() -> Target {
```
Comment (cardoe) [inline src/librustc_back/target/armv7_unknown_linux_musleabihf.rs]: This needs to return `TargetResult` now.

**Comment by japaric** [comment]
> This needs to return TargetResult now.

Thanks.

> Any reason why the GNU Make files aren't being populated and we have to use the new build system to build these targets?

Probably it just hasn't been tested.

@bors r=alexcrichton rollup

**Pull request merged**